### ISSUE TITLE
[ActiveRecord] InsertAll ignores integer timestamp

### ISSUE THREAD

**Issue by davidpiegza** [rails/rails#53317]
After [this change in `InsertAll`](https://github.com/rails/rails/pull/48139/files) the value of a custom `DateTime` attribute is not saved when using an integer timestamp.

This happens in a specific case when using a custom timestamp attribute, which should save the value as an integer timestamp but behave like a `DateTime` column in Rails.

### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  
  # gem "rails"
  
  # If you want to test against edge Rails replace the previous line with this:
  gem "rails", github: "rails/rails", branch: "main"
  
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Base.time_zone_aware_attributes = true
ActiveRecord.default_timezone = :local

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.bigint :expires_at
  end
end

class EpochTimestamp < ActiveRecord::Type::DateTime
  def deserialize(time_or_int)
    Time.at(time_or_int).utc if time_or_int
  end
  
  def serialize(time)
    time.to_i if time
  end
end

ActiveRecord::Type.register(:epoch_timestamp, EpochTimestamp)

class Post < ActiveRecord::Base
  attribute :expires_at, :epoch_timestamp
end

class BugTest < ActiveSupport::TestCase
  def test_custom_timestamp_with_insert_all
    timestamp = Time.current.utc
    
    Post.insert_all!([{expires_at: timestamp.to_i}]) # this will not save the integer timestamp
    
    assert_equal timestamp.to_i, Post.last.expires_at.to_i # Post.last.expires_at == nil
  end
end
```

I created this issue to discuss if this is a potential breaking change for more users and if it can be fixed upstream. See also [this comment](https://github.com/rails/rails/pull/48139#issuecomment-2413566763). 

cc @OuYangJinTing as author of the change.

**Comment by matthewd** [comment]
@kamipo I can't decide whether this is a breaking change we need to be concerned about, or just evidence that we-GitHub were previously misusing the API in a way no-one else is likely to.

#48139 is for sure the more correct behaviour... it might just be too hard to provide a more gradual transition, even if we wanted to? 🤔

**Comment by kamipo** [comment]
I've identified it is a potential bug of `map_avoiding_infinite_recursion` introduced at 5bb26008ce7c979a87f0d8aef0dd10514838787f.
I'll make a fix soon.

**Comment by matthewd** [comment]
Ah, I hadn't considered that it might be something specific to the fact it's a DateTime value -- interesting! 🤔 

I think my general concern is whether there are going to be other instances of the more broad situation, which is just "application is currently providing a value that [works when serialized, but] converts to a different stored value when we `cast` it first". Our integer-into-nil case is extreme, but I wonder if there could be others.

**Comment by kamipo** [comment]
I've made a fix #53359.

I see your concern, so I checked all attribute types which are provided by us.

https://github.com/rails/rails/blob/e71a6d410e57589fc9e4b13c918eed427c936152/activemodel/lib/active_model/type/value.rb#L37-L59
https://github.com/rails/rails/blob/e71a6d410e57589fc9e4b13c918eed427c936152/activemodel/lib/active_model/type/value.rb#L149-L154

Since `Type::Value#deserialize` method is defined as `def deserialize(value) = cast(value)`, almost all our provided types define `cast` or `cast_value`

(except `OID::Bytea` (use `unescape_bytea` to the raw input from database), and types that include `Mutable` helper such as `Json` (`Mutable` helper provides `def cast(value) = deserialize(serialize(value))`, so requires `deserialize` or `cast` over-ridden),

so when valid serialized values are given, all our provided both `cast` and `deserialize` can properly type-cast that values to Ruby world's objects.

The integer-into-nil case is a bug which will be fixed by #53359. At least it is possible to override only `deserialize` (but borrow `cast` from the original type) to make it behave asymmetrically as `cast`, but if any users create an asymmetry that makes it impossible to cast valid serialized values, `Post.insert!({ expires_at: serialized_value })` does work but `Post.create!(expires_at: serialized_value)` does not.

So I don't suppose there are any users who are using custom types that cannot cast valid serialized values, but we could conservatively fall back to the original value.

```diff
diff --git a/activerecord/lib/active_record/insert_all.rb b/activerecord/lib/active_record/insert_all.rb
index 66a6c3204b..82bff4212d 100644
--- a/activerecord/lib/active_record/insert_all.rb
+++ b/activerecord/lib/active_record/insert_all.rb
@@ -240,7 +240,7 @@ def values_list
 
           values_list = insert_all.map_key_with_value do |key, value|
             next value if Arel::Nodes::SqlLiteral === value
-            types[key].serialize(types[key].cast(value))
+            types[key].serialize(types[key].cast(value) || value)
           end
 
           connection.visitor.compile(Arel::Nodes::ValuesList.new(values_list))
```

Do you think we should do that?

**Comment by matthewd** [comment]
> `Post.insert!({ expires_at: serialized_value })` does work but `Post.create!(expires_at: serialized_value)` does not

Yeah.. I was definitely surprised to realise that that was our situation -- I assume our call sites had added the `to_i` specifically as a work-around for the previously-missing `cast`.

> we could conservatively fall back to the original value

That would have helped in this particular situation, because the extra `cast` was producing a `nil`, but I don't think that would be the general case.

Agree that [apart from identified #53359 🙇🏻‍♂️] our internal types would expect to work with the addition of a "bonus" `cast`, even where (as in our example `to_i` situation) the current callsite has implemented the first cast manually... and that's probably the best we can do.

> Do you think we should do that?

No, thanks for talking it through with me. I think #48139 is doing the right thing as is. There is some chance that someone might have an arguably-broken custom Type that would have a problem here during an upgrade, but I don't think there's anything we can do about that beyond apologising (👋🏻 future readers 😅), and recommending that `cast` should generally work for an already-cast value.
